src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingObjectMatchProcessor.cs

```diff
@@ -0,0 +1,131 @@
+using Microsoft.SqlTools.ServiceLayer.Scripting.Contracts;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Scripting
+{
+    /// <summary>
+    /// Implementes the matchin logic to filter scripting objects based on
+    /// an include/exclude criteria.
+    /// </summary>
+    /// <remarks>
+    /// First, objects are included by the include filter.  Then, objects are removed by
+    /// the exclude filter.  Matches are made by comparing case insensitive strings.  
+    /// Wildcard '*' are supported for all scripting object fields.
+    /// </remarks>
+    public static class ScriptingObjectMatchProcessor
+    {
+        private const string Wildcard = "*";
+
+        /// <summary>
+        /// Given a collection of candidate scripting objects, filters the items that match 
+        /// based on the passed include and exclude criteria.
+        /// </summary>
+        /// <param name="includeCriteria">The include object criteria.</param>
+        /// <param name="excludeCriteria">The exclude object criteria.</param>
+        /// <param name="candidates">The candidate object to filter.</param>
+        /// <returns>The matching scripting objects.</returns>
+        public static IEnumerable<ScriptingObject> Match(
+            ScriptingObject includeCriteria,
+            ScriptingObject excludeCriteria,
+            IEnumerable<ScriptingObject> candidates)
+        {
+            return Match(
+                includeCriteria == null ? new ScriptingObject[0] : new[] { includeCriteria },
+                excludeCriteria == null ? new ScriptingObject[0] : new[] { excludeCriteria },
+                candidates);
+        }
+
+        /// <summary>
+        /// Given a collection of candidate scripting objects, filters the items that match 
+        /// based on the passed include and exclude criteria.
+        /// </summary>
+        /// <param name="includeCriteria">The collection of include object criteria items.</param>
+        /// <param name="excludeCriteria">The collection of exclude object criteria items.</param>
+        /// <param name="candidates">The candidate object to filter.</param>
+        /// <returns>The matching scripting objects.</returns>
+        public static IEnumerable<ScriptingObject> Match(
+            IEnumerable<ScriptingObject> includeCriteria,
+            IEnumerable<ScriptingObject> excludeCriteria,
+            IEnumerable<ScriptingObject> candidates)
+        {
+            Validate.IsNotNull("candidates", candidates);
+
+            IEnumerable<ScriptingObject> matchedObjects = new List<ScriptingObject>();
+
+            if (includeCriteria != null && includeCriteria.Any())
+            {
+                foreach (ScriptingObject scriptingObjectCriteria in includeCriteria)
+                {
+                    IEnumerable<ScriptingObject> matches = MatchCriteria(scriptingObjectCriteria, candidates);
+                    matchedObjects = matchedObjects.Union(matches);
```

Comment:
No craziness for me. To be frank, I don't even know how I would accomplish that with an aggregate.